### PR TITLE
Auto Delete Empty Anchors and Sources in SQL Registry

### LINKED ISSUE #1022: [FR] Empty Anchor and Source could be auto deleted.
### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### Feature Request Proposal

Current UI doesn't have "anchor" or "source" page to delete those entities, which blocks project deletion in UI.

### Motivation

> #### What is the use case for this feature?
Delete a whole project rather than several features.

### Details

_No response_

### What component(s) does this feature request affect?

- [ ] `Python Client`: This is the client users use to interact with most of our API. Mostly written in Python.
- [ ] `Computation Engine`: The computation engine that execute the actual feature join and generation work. Mostly in Scala and Spark.
- [X] `Feature Registry API`: The frontend API layer supports SQL, Purview(Atlas) as storage. The API layer is in Python(FAST API)
- [ ] `Feature Registry Web UI`: The Web UI for feature registry. Written in React

### PR DESCRIPTION
**Pull request by Yuqing-cat**
Signed-off-by: Yuqing Wei <weiyuqing021@outlook.com>

## Description

Resolves #1022 
Add a `delete_empty_entities ` function to clean up empty anchors and sources in given entity list. 

## How was this PR tested?

Project in SQL registry can be deleted successfully after all features are removed. 
image
image

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to clarify your proposed changes.

### CODE REVIEW

File: registry/sql-registry/registry/db_registry.py [hunk 1]
```diff
@@ -204,6 +204,28 @@ def get_dependent_entities(self, entity_id: Union[str, UUID]) -> List[Entity]:
...
+        for e in entities:
+            if e.entity_type == EntityType.Anchor:
+                downstream_entities, _ = self._bfs(e.id, RelationshipType.Contains) 
+                if len(downstream_entities) == 1: # only anchor itself
+                    self.delete_entity(e.id)
+        # clean up empty sources
+        for e in entities:
+            if e.entity_type == EntityType.Source:
+                downstream_entities, _ = self._bfs(e.id, RelationshipType.Produces) 
+                if len(downstream_entities) == 1: # only source itself
```
Comment (xiaoyongzhu) [inline registry/sql-registry/registry/db_registry.py:224]: Do we need to check the returned entity type? Just to make sure downstream_entities is Anchor/Source?

Comment (Yuqing-cat) [reply registry/sql-registry/registry/db_registry.py:224]: the _bfs entity list return will always contain the entity itself (`e`), which type is already checked above (`e.entity_type`).

Comment (enya-yx) [reply registry/sql-registry/registry/db_registry.py:224]: Is that better we use one for loop here? Eg. for e in entities: if type is Anchor: ...; elif type is Source: ...

Comment (Yuqing-cat) [reply registry/sql-registry/registry/db_registry.py:224]: > Is that better we use one for loop here? Eg. for e in entities: if type is Anchor: ...; elif type is Source: ...

`Source` produces `Anchor`, they have sequence.  

If merged in one loop, for `Source`, you need to check if all the downstream entities has empty `Anchor` which also involves another loop.

But if all empty anchors are cleaned in the first loop, then we can clean empty sources directly.

**Review by xiaoyongzhu — APPROVED** [review]

File: registry/sql-registry/registry/db_registry.py (hunk 1 above)
Comment (enya-yx) [reply registry/sql-registry/registry/db_registry.py:224]: > > Is that better we use one for loop here? Eg. for e in entities: if type is Anchor: ...; elif type is Source: ...
> 
> `Source` produces `Anchor`, they have sequence.
> 
> If merged in one loop, for `Source`, you need to check if all the downstream entities has empty `Anchor` which also involves another loop.
> 
> But if all empty anchors are cleaned in the first loop, then we can clean empty sources directly.

Good point

**Review by enya-yx — APPROVED** [review]

**Pull request merged**